warp10/src/main/java/io/warp10/script/functions/HHCODETO.java

```diff
@@ -44,34 +48,56 @@ public HHCODETO(String name, boolean useGtsConvention) {
   @Override
   public Object apply(WarpScriptStack stack) throws WarpScriptException {
 
-    Object hhcode = stack.pop();
+    Object top = stack.pop();
+
+    if (top instanceof List) {
+      List cellList = (List) top;
+      ArrayList<Long> cells = new ArrayList<Long>();
 
-    long hh;
+      long[] hhAndRes;
+      for(Object cell: cellList) {
+        try {
+          hhAndRes = HHCODEFUNC.hhAndRes(cell);
+        } catch (WarpScriptException wse) {
+          throw new WarpScriptException(getName() + " expects the given list to contain LONG, STRING or BYTES HHCodes.", wse);
+        }
+        long hh = hhAndRes[0];
+        int res = (int) hhAndRes[1]; // We know hhAndRes returns an int here.
 
-    if (hhcode instanceof Long) {
-      hh = (long) hhcode;
-    } else if (hhcode instanceof String) {
-      String hhstr = hhcode.toString();
-      if (hhstr.length() > 16) {
-        throw new WarpScriptException(getName() + " expects an hexadecimal HHCode string of length <= 16");
-      } else if (hhstr.length() < 16) {
-        hhcode = new StringBuilder(hhstr).append("0000000000000000");
-        ((StringBuilder) hhcode).setLength(16);
+        cells.add(HHCodeHelper.toGeoCell(hh, res));
       }
-      hh = Long.parseUnsignedLong(hhcode.toString(), 16);
-    } else if (hhcode instanceof byte[]) {
-      hh = Longs.fromByteArray((byte[]) hhcode);
-    } else {
-      throw new WarpScriptException(getName() + " expects a long, a string or a byte array.");
-    }
 
-    if (useGtsConvention && GeoTimeSerie.NO_LOCATION == hh) {
-      stack.push(Double.NaN);
-      stack.push(Double.NaN);
+      stack.push(GeoXPLib.fromCells(cells));
     } else {
-      double[] latlon = GeoXPLib.fromGeoXPPoint(hh);
-      stack.push(latlon[0]);
-      stack.push(latlon[1]);
+      // Optional: boolean to select GeoHash/Lat-Lon output
+      boolean geohashOutput = false;
+      if (top instanceof Boolean) {
+        geohashOutput = (Boolean) top;
+        top = stack.pop();
+      }
+
+      long[] hhAndRes;
+      try {
+        hhAndRes = HHCODEFUNC.hhAndRes(top);
+      } catch (WarpScriptException wse) {
+        throw new WarpScriptException(getName() + " was given unexpected arguments.", wse);
+      }
+      long hh = hhAndRes[0];
+      int res = (int) hhAndRes[1]; // We know hhAndRes returns an int here.
+
+      if(geohashOutput) {
+        // This is an approximation of the HHCode with a GeoHash
+        stack.push(GeoHashHelper.fromHHCode(hh, res));
+      } else {
+        if (useGtsConvention && GeoTimeSerie.NO_LOCATION == hh) {
+          stack.push(Double.NaN);
+          stack.push(Double.NaN);
+        } else {
+          double[] latlon = HHCodeHelper.getCenterLatLon(hh, res);
```

Comment:
This violates the current behavior. A HHCode encodes the lower left corner of its cell, not the cell center.